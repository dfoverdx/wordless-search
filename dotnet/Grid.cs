using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace WordlessSearch
{
    using Point = Tuple<int, int>;

    public partial class Grid
    {
        private char[,] grid;

        private bool gridDirty = true;
        private readonly char[,] _gridT;
        private char[,] gridT
        {
            get {
                if (gridDirty)
                {
                    for (int y = 0; y < Size; y++)
                    {
                        for (int x = 0; x < Size; x++)
                        {
                            _gridT[x, y] = grid[y, x];
                        }
                    }

                    gridDirty = false;
                }

                return _gridT;
            }
        }

        private readonly int Size;

        private readonly WordPos[] StaticWords;

        private readonly HashSet<WordPos> StaticWordFragments;

        private readonly static Random Random = new Random();

        public bool Verbose = true;

        public Grid(int size = Constants.GridSize)
        {
            Size = size;
            grid = new char[size, size];
            _gridT = new char[size, size];
            charBuffer = new char[Size];

            for (int y = 0; y < Size; y++)
            {
                for (int x = 0; x < Size; x++)
                {
                    grid[y, x] = ' ';
                }
            }

            StaticWords = new WordPos[] {
                new WordPos { Word = "FRIENDS", Direction = Direction.East, Point = new Point(1, 1) },
                new WordPos { Word = "MONEY", Direction = Direction.South, Point = new Point(size - 1, 2) }
            };

            SetStaticWords();

            StaticWordFragments = new HashSet<WordPos>();
            StaticWordFragments = FindWords().ToHashSet();

            Reset();
            SetStaticWords();
        }

        public void Reset()
        {
            gridDirty = true;
            for (int y = 0; y < Size; y++)
            {
                for (int x = 0; x < Size; x++)
                {
                    grid[y, x] = Words.RandomLetter();
                }
            }
        }

        private void SetStaticWords()
        {
            foreach (WordPos word in StaticWords)
            {
                SetWord(word);
            }
        }

        public void DoEvil()
        {
            if (Verbose)
            {
                Console.Clear();
            }

            while (FixWords() | FixVowels() | FixRuns());
        }
    }

    public static class PointExtensions
    {
        public static Point Move(this Point point, Direction direction, int distance = 1)
        {
            var (x, y) = point;

            switch (direction)
            {
                case Direction.North:
                case Direction.Northeast:
                case Direction.Northwest:
                    y -= distance;
                    break;

                case Direction.South:
                case Direction.Southeast:
                case Direction.Southwest:
                    y += distance;
                    break;
            }

            switch (direction)
            {
                case Direction.East:
                case Direction.Northeast:
                case Direction.Southeast:
                    x += distance;
                    break;

                case Direction.West:
                case Direction.Northwest:
                case Direction.Southwest:
                    x -= distance;
                    break;
            }

            return new Point(x, y);
        }
    }

    public enum Direction
    {
        North,
        Northwest,
        West,
        Southwest,
        South,
        Southeast,
        East,
        Northeast
    }

    public struct WordPos
    {
        public string Word;
        public int Length => Word.Length;
        public Direction Direction;
        public Point Point;

        public static bool operator==(WordPos left, WordPos right)
        {
            return left.Equals(right);
        }

        public static bool operator!=(WordPos left, WordPos right)
        {
            return !left.Equals(right);
        }

        public override bool Equals(object other)
        {
            if (!(other is WordPos))
            {
                return false;
            }

            WordPos _other = (WordPos)other;
            return Direction == _other.Direction &&
                Word == _other.Word &&
                Point.Item1 == _other.Point.Item1 &&
                Point.Item2 == _other.Point.Item2;
        }

        public override int GetHashCode()
        {
            return Word.GetHashCode() * Direction.GetHashCode() * Point.GetHashCode();
        }
    }
}