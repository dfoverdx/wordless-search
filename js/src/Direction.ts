enum Direction {
  North = 'North',
  Northwest = 'Northwest',
  West = 'West',
  Southwest = 'Southwest',
  South = 'South',
  Southeast = 'Southeast',
  East = 'East',
  Northeast = 'Northeast',
}

export default Direction;