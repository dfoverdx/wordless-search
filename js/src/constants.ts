export const Size = 16;
export const NumAllowedWords = 2;
// export const AllowedWordMinLength = 5;
export const MinVowelRatio = 1 / 10;

export const MinVowels = Size * Size * MinVowelRatio;