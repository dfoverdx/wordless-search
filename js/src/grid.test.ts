import { Size } from './constants';
import Direction from './Direction';
import Grid from './grid';

const GridConstructor = Grid.constructor as new (size: number) => typeof Grid;

describe('Grid', () => {
  let grid: typeof Grid;

  beforeEach(() => {
    Grid.verbose = false;
    Grid.highlightWords = false;

    const testVal = `
      CATS
      ERAY
      ETAG
      FYAP
    `.replace(/ +/g, '').trim().split('\n').map(row => row.split(''));
    grid = new GridConstructor(4);
    grid[0] = testVal[0];
    grid[1] = testVal[1];
    grid[2] = testVal[2];
    grid[3] = testVal[3];
  });

  afterEach(jest.resetAllMocks);

  it(`is ${Size}Ã—${Size} in length`, () => {
    expect(Grid.length).toBe(Size);
    expect(Grid.every(row => row.length === Size)).toBe(true);
  });

  it('is filled with single-character strings', () => {
    expect(Grid.every(row => row.every(cell => cell.length === 1))).toBe(true);
  });

  it('is filled with capital letters', () => {
    const str = Grid.map(row => row.join('')).join('');
    expect(str).toBe(str.toUpperCase());
  });

  describe('toString()', () => {
    it('spaces characters properly', () => {
      expect(Grid.toString()).toMatch(new RegExp(`^(([A-Z] ){${Size - 1}}[A-Z]\n\n){${Size - 1}}([A-Z] ){${Size - 1}}[A-Z]$`));
    });

    it('colors things properly', () => {
      grid.highlightWords = true;

    });
  });

  describe('findWords()', () => {
    it('finds the correct words', () => {
      expect(grid.findWords()).toMatchSnapshot();
    });
  });

  describe('breakWord()', () => {
    it('replaces a random letter within the word a different letter', () => {
      const spy = jest.spyOn(Math, 'randomInt');
      spy.mockReturnValueOnce(1);
      spy.mockReturnValueOnce(65);
      spy.mockReturnValueOnce(66);

      grid.breakWord({
        x: 0,
        y: 0,
        direction: Direction.East,
        length: 3,
        word: 'CAT',
      });

      expect(grid[0]).toEqual('CBTS'.split(''));
    });
  });
});