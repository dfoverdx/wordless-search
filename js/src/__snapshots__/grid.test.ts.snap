// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid findWords() finds the correct words 1`] = `
Array [
  Object {
    "direction": "South",
    "length": 3,
    "word": "CEE",
    "x": 0,
    "y": 0,
  },
  Object {
    "direction": "Southeast",
    "length": 4,
    "word": "CRAP",
    "x": 0,
    "y": 0,
  },
  Object {
    "direction": "East",
    "length": 3,
    "word": "CAT",
    "x": 0,
    "y": 0,
  },
  Object {
    "direction": "Southeast",
    "length": 3,
    "word": "ETA",
    "x": 0,
    "y": 1,
  },
  Object {
    "direction": "East",
    "length": 3,
    "word": "ERA",
    "x": 0,
    "y": 1,
  },
  Object {
    "direction": "East",
    "length": 3,
    "word": "ETA",
    "x": 0,
    "y": 2,
  },
  Object {
    "direction": "North",
    "length": 3,
    "word": "FEE",
    "x": 0,
    "y": 3,
  },
  Object {
    "direction": "South",
    "length": 3,
    "word": "ART",
    "x": 1,
    "y": 0,
  },
  Object {
    "direction": "East",
    "length": 3,
    "word": "ATS",
    "x": 1,
    "y": 0,
  },
  Object {
    "direction": "Southeast",
    "length": 3,
    "word": "RAP",
    "x": 1,
    "y": 1,
  },
  Object {
    "direction": "East",
    "length": 3,
    "word": "RAY",
    "x": 1,
    "y": 1,
  },
  Object {
    "direction": "East",
    "length": 3,
    "word": "TAG",
    "x": 1,
    "y": 2,
  },
  Object {
    "direction": "Northeast",
    "length": 3,
    "word": "TAS",
    "x": 1,
    "y": 2,
  },
  Object {
    "direction": "East",
    "length": 3,
    "word": "YAP",
    "x": 1,
    "y": 3,
  },
  Object {
    "direction": "Northeast",
    "length": 3,
    "word": "YAY",
    "x": 1,
    "y": 3,
  },
  Object {
    "direction": "West",
    "length": 3,
    "word": "ARE",
    "x": 2,
    "y": 1,
  },
  Object {
    "direction": "Northwest",
    "length": 3,
    "word": "ARC",
    "x": 2,
    "y": 2,
  },
  Object {
    "direction": "West",
    "length": 3,
    "word": "ATE",
    "x": 2,
    "y": 2,
  },
  Object {
    "direction": "Northwest",
    "length": 3,
    "word": "ATE",
    "x": 2,
    "y": 3,
  },
  Object {
    "direction": "Southwest",
    "length": 3,
    "word": "SAT",
    "x": 3,
    "y": 0,
  },
  Object {
    "direction": "West",
    "length": 3,
    "word": "YAR",
    "x": 3,
    "y": 1,
  },
  Object {
    "direction": "Southwest",
    "length": 3,
    "word": "YAY",
    "x": 3,
    "y": 1,
  },
  Object {
    "direction": "West",
    "length": 3,
    "word": "GAT",
    "x": 3,
    "y": 2,
  },
  Object {
    "direction": "Northwest",
    "length": 3,
    "word": "PAR",
    "x": 3,
    "y": 3,
  },
  Object {
    "direction": "West",
    "length": 3,
    "word": "PAY",
    "x": 3,
    "y": 3,
  },
]
`;
